File: Hard/2025-10-12-3539-Find-Sum-of-Array-Product-of-Magical-Sequences/solution.js
class Solution {
    static MOD = 1e9 + 7;

    findSum(m, k, nums) {
        const n = nums.length;
        const dp = Array.from({ length: m + 1 }, () => Array(1 << n).fill(0));
        dp[0][0] = 1;

        for (let i = 0; i < m; i++) {
            for (let mask = 0; mask < (1 << n); mask++) {
                if (dp[i][mask] === 0) continue;
                for (let j = 0; j < n; j++) {
                    const newMask = mask | (1 << j);
                    dp[i + 1][newMask] = (dp[i + 1][newMask] + dp[i][mask]) % Solution.MOD;
                }
            }
        }

        let result = 0;
        for (let mask = 0; mask < (1 << n); mask++) {
            if (this.countSetBits(mask) === k) {
                let product = 1;
                for (let j = 0; j < n; j++) {
                    if (mask & (1 << j)) {
                        product = (product * nums[j]) % Solution.MOD;
                    }
                }
                result = (result + product) % Solution.MOD;
            }
        }

        return result;
    }

    countSetBits(num) {
        let count = 0;
        while (num > 0) {
            count += num & 1;
            num >>= 1;
        }
        return count;
    }
}

File: Hard/2025-10-12-3539-Find-Sum-of-Array-Product-of-Magical-Sequences/solution.java
import java.util.*;

public class Solution {
    private static final int MOD = 1000000007;

    public int findSumOfArrayProduct(int m, int k, int[] nums) {
        int n = nums.length;
        long[][] dp = new long[m + 1][k + 1];
        dp[0][0] = 1;

        for (int i = 0; i < n; i++) {
            for (int j = m; j >= 1; j--) {
                for (int l = k; l >= 1; l--) {
                    dp[j][l] = (dp[j][l] + dp[j - 1][l - 1] * nums[i]) % MOD;
                }
            }
        }

        long result = 0;
        for (int j = 1; j <= m; j++) {
            result = (result + dp[j][k]) % MOD;
        }

        return (int) result;
    }
}